server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileRef.java

```diff
@@ -0,0 +1,155 @@
+    this.key = key;
+    this.language = language;
+    this.name = name;
+  }
+
+  /**
+   * @return {@code true} if key is defined and {@link #getKey()} can be called. If {@code false}, then
+   *   the couple {language, name} is defined and the methods {@link #getLanguage()}/{@link #getName()}
+   *   can be called.
+   */
+  public boolean hasKey() {
+    return this.key != null;
+  }
+
+  /**
+   * @return non-null key
+   * @throws IllegalStateException if {@link #hasKey()} does not return {@code true}
+   */
+  public String getKey() {
+    checkState(key != null, "Key is not present. Please call hasKey().");
+    return key;
+  }
+
+  /**
+   * @return non-null language
+   * @throws IllegalStateException if {@link #hasKey()} does not return {@code false}
+   */
+  public String getLanguage() {
+    checkState(language != null, "Language is not present. Please call hasKey().");
+    return language;
+  }
+
+  /**
+   * @return non-null name
+   * @throws IllegalStateException if {@link #hasKey()} does not return {@code false}
+   */
+  public String getName() {
+    checkState(name != null, "Name is not present. Please call hasKey().");
+    return name;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    QProfileRef that = (QProfileRef) o;
+    if (key != null ? !key.equals(that.key) : that.key != null) {
+      return false;
+    }
+    if (language != null ? !language.equals(that.language) : that.language != null) {
+      return false;
+    }
+    return name != null ? name.equals(that.name) : that.name == null;
+
+  }
+
+  @Override
+  public int hashCode() {
+    int result = key != null ? key.hashCode() : 0;
+    result = 31 * result + (language != null ? language.hashCode() : 0);
+    result = 31 * result + (name != null ? name.hashCode() : 0);
+    return result;
+  }
+
+  public static QProfileRef from(Request request) {
+    String key = request.param(PARAM_PROFILE_KEY);
+    String lang = request.param(PARAM_LANGUAGE);
+    String name = request.param(PARAM_PROFILE_NAME);
+    return from(key, lang, name);
+  }
+
+  public static QProfileRef from(@Nullable String key, @Nullable String lang, @Nullable String name) {
+    if (key != null) {
+      checkArgument(isEmpty(lang) && isEmpty(name), "Either key or couple language/name must be set");
+      return fromKey(key);
+    }
+    checkArgument(!isEmpty(lang) && !isEmpty(name), "Both profile language and name must be set");
+    return fromName(lang, name);
+  }
+
+  public static QProfileRef fromKey(String key) {
+    return new QProfileRef(requireNonNull(key), null, null);
+  }
+
+  public static QProfileRef fromName(String lang, String name) {
+    return new QProfileRef(null, requireNonNull(lang), requireNonNull(name));
+  }
+
+  public static void defineParams(WebService.NewAction action, Languages languages) {
```

Comment:
I would not put this method here, but we've already talked about it : ) As this method is now static (so no need of Pico component), I would put it with parameters in a `QualityProfilesWsParameters` that what's done in other domains. I'd be interested to know why it's a bad idea so we can change for future domains